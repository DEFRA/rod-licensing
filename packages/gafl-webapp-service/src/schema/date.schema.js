'use strict'
import Joi from 'joi'

export const dateSchemaInput = (day, month, year) => ({
  'full-date': { day, month, year },
  'day-and-month': { day, month },
  'day-and-year': { day, year },
  'month-and-year': { month, year },
  day,
  month,
  year,
  'non-numeric': { day, month, year },
  'invalid-date': `${year}-${(month || '').padStart(2, '0')}-${(day || '').padStart(2, '0')}`
})

export const dateSchema = Joi.object({
  'full-date': Joi.object()
    .keys({
      day: Joi.any(),
      month: Joi.any(),
      year: Joi.any()
    })
    .or('day', 'month', 'year'),
  'day-and-month': Joi.object()
    .keys({
      day: Joi.any(),
      month: Joi.any()
    })
    .or('day', 'month'),
  'day-and-year': Joi.object()
    .keys({
      day: Joi.any(),
      year: Joi.any()
    })
    .or('day', 'year'),
  'month-and-year': Joi.object()
    .keys({
      month: Joi.any(),
      year: Joi.any()
    })
    .or('month', 'year'),
  day: Joi.any().required(),
  month: Joi.any().required(),
  year: Joi.any().required(),
  'non-numeric': Joi.object().keys({
    day: Joi.number(),
    month: Joi.number(),
    year: Joi.number()
  }),
  // 'invalid-date': Joi.date().iso().strict()
  'invalid-date': Joi.custom((dateToValidate, helpers) => {
    if ((new Date(dateToValidate)).toISOString() !== `${dateToValidate}T00:00:00.000Z`) {
      throw helpers.error('invalid-date')
    }

    return dateToValidate
  })
}).options({ abortEarly: true })
