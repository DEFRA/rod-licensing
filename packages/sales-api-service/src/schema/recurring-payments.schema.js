import Joi from 'joi'
import { validation } from '@defra-fish/business-rules-lib'

export const recurringPaymentsRequestParamsSchema = Joi.object({
  date: validation.date.isoDate().description('The date for which recurring payments are being retrieved')
}).label('recurring-payments-request-params')

const recurringPaymentSchema = Joi.object({
  id: Joi.string().required().description('The ID of the recurring payment'),
  amount: Joi.number().required().description('The payment amount of the recurring payment'),
  dueDate: validation.date.isoDate().description('The due date of the recurring payment'),
  status: Joi.string().valid('pending', 'paid', 'failed').required().description('The status of the payment')
}).label('recurring-payment')

export const recurringPaymentsResponseSchema = Joi.array()
  .items(recurringPaymentSchema)
  .label('recurring-payments-response')
