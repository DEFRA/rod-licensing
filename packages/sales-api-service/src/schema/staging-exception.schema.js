import Joi from 'joi'
import { buildJoiOptionSetValidator, createAlternateKeyValidator } from './validators/validators.js'
import { PoclFile, PoclStagingException } from '@defra-fish/dynamics-lib'

const schemaObject = {
  stagingException: Joi.object({
    stagingId: Joi.string().required(),
    description: Joi.string().required(),
    transactionJson: Joi.string().required(),
    exceptionJson: Joi.string().required()
  }),
  transactionFileException: Joi.object({
    name: Joi.string().required(),
    description: Joi.string(),
    json: Joi.string().required(),
    notes: Joi.string(),
    type: buildJoiOptionSetValidator(PoclStagingException.definition.mappings.type.ref, 'Failure'),
    transactionFile: Joi.string()
      .external(createAlternateKeyValidator(PoclFile))
      .required(),
    permissionId: Joi.string()
  }),
  record: Joi.object({
    id: Joi.string(),
    createTransactionPayload: Joi.object({
      dataSource: Joi.string(),
      serialNumber: Joi.string(),
      permissions: Joi.array()
    }),
    finaliseTransactionPayload: Joi.object({
      payment: Joi.object({
        timestamp: Joi.string(),
        amount: Joi.number(),
        source: Joi.string(),
        channelId: Joi.string(),
        method: Joi.string()
      })
    }),
    stage: Joi.string(),
    createTransactionError: Joi.object({
      statusCode: Joi.number(),
      error: Joi.string(),
      message: Joi.string()
    })
  }).optional()
}

/**
 * Schema for the create staging exception request
 * @type {Joi.AnySchema}
 */
export const createStagingExceptionRequestSchema = Joi.object(schemaObject)
  .or('stagingException', 'transactionFileException')
  .label('create-staging-exception-request')

/**
 * Schema for the create staging exception response
 * @type {Joi.AnySchema}
 */
export const createStagingExceptionResponseSchema = Joi.object(schemaObject).label('create-staging-exception-response')
