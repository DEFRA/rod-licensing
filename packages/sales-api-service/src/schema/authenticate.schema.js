import Joi from 'joi'
import { validation } from '@defra-fish/business-rules-lib'
import { concessionProofSchema } from './concession-proof.schema.js'
import { permitSchema } from './permit.schema.js'
import { contactResponseSchema } from './contact.schema.js'
import { finalisedPermissionSchemaContent } from './permission.schema.js'

export const authenticateRenewalRequestParamsSchema = Joi.object({
  referenceNumber: Joi.string().min(6).required().description('The permission reference number (supports partial)')
}).label('authenticate-renewal-request-params')

export const authenticateRenewalRequestQuerySchema = Joi.object({
  licenseeBirthDate: validation.contact.createBirthDateValidator(Joi).description('The date of birth of the licensee'),
  licenseePostcode: Joi.alternatives().try(
    validation.contact.createUKPostcodeValidator(Joi).description('The postcode of the licensee'),
    validation.contact.createOverseasPostcodeValidator(Joi)
  )
}).label('authenticate-renewal-request-query')

export const authenticateRenewalResponseSchema = Joi.object({
  permission: {
    ...finalisedPermissionSchemaContent,
    licensee: contactResponseSchema,
    concessions: concessionProofSchema,
    permit: permitSchema
  }
}).label('authenticate-renewal-response')
