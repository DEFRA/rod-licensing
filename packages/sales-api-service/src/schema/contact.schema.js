import Joi from 'joi'
import { buildJoiOptionSetValidator, createEntityIdValidator } from './validators/validators.js'
import { Contact } from '@defra-fish/dynamics-lib'
import { validation, POCL_TRANSACTION_SOURCES } from '@defra-fish/business-rules-lib'
import { optionSetOption } from './option-set.schema.js'
const DATASOURCE_REF = '/dataSource'

const commonContactSchema = {
  id: Joi.string()
    .guid()
    .optional()
    .external(createEntityIdValidator(Contact))
    .description('the contact identifier of an existing contact record to be updated')
    .example('1329a866-d175-ea11-a811-000d3a64905b'),
  firstName: validation.contact.createFirstNameValidator(Joi, { minimumLength: 1 }),
  lastName: validation.contact.createLastNameValidator(Joi, { minimumLength: 1 }),
  birthDate: validation.contact.createBirthDateValidator(Joi),
  email: validation.contact.createEmailValidator(Joi).allow(null),
  mobilePhone: validation.contact.createMobilePhoneValidator(Joi).allow(null),
  organisation: Joi.string()
    .trim()
    .min(1)
    .optional()
    .allow(null)
    .example('Example Organisation'),
  premises: Joi.when(Joi.ref(DATASOURCE_REF), {
    is: Joi.string().valid(...POCL_TRANSACTION_SOURCES),
    then: validation.contact
      .createPremisesValidator(Joi)
      .optional()
      .allow(null, ''),
    otherwise: validation.contact.createPremisesValidator(Joi)
  }).example('Example House'),
  street: validation.contact.createStreetValidator(Joi).allow(null),
  locality: validation.contact.createLocalityValidator(Joi).allow(null),
  town: Joi.when(Joi.ref(DATASOURCE_REF), {
    is: Joi.string().valid(...POCL_TRANSACTION_SOURCES),
    then: validation.contact
      .createTownValidator(Joi)
      .optional()
      .allow(null, ''),
    otherwise: validation.contact.createTownValidator(Joi)
  }).example('Exampleton'),
  postcode: Joi.when(Joi.ref(DATASOURCE_REF), {
    is: Joi.string().valid(...POCL_TRANSACTION_SOURCES),
    then: Joi.alternatives().conditional('country', {
      is: Joi.string().valid('GB', 'United Kingdom'),
      then: validation.contact
        .createUKPostcodeValidator(Joi)
        .optional()
        .allow(null, ''),
      otherwise: validation.contact
        .createOverseasPostcodeValidator(Joi)
        .optional()
        .allow(null, '')
    }),
    otherwise: Joi.alternatives().conditional('country', {
      is: Joi.string().valid('GB', 'United Kingdom'),
      then: validation.contact.createUKPostcodeValidator(Joi),
      otherwise: validation.contact.createOverseasPostcodeValidator(Joi)
    })
  }).example('AB12 3CD'),
  postalFulfilment: Joi.boolean()
    .optional()
    .allow(null)
    .example('true')
}

export const contactRequestSchema = Joi.object({
  ...commonContactSchema,
  country: buildJoiOptionSetValidator('defra_country', 'GB'),
  preferredMethodOfConfirmation: buildJoiOptionSetValidator('defra_preferredcontactmethod', 'Text'),
  preferredMethodOfNewsletter: buildJoiOptionSetValidator('defra_preferredcontactmethod', 'Email'),
  preferredMethodOfReminder: buildJoiOptionSetValidator('defra_preferredcontactmethod', 'Letter')
})
  .required()
  .description('Details of the associated contact')
  .label('contact')

export const contactResponseSchema = Joi.object({
  ...commonContactSchema,
  country: optionSetOption,
  preferredMethodOfConfirmation: optionSetOption,
  preferredMethodOfNewsletter: optionSetOption,
  preferredMethodOfReminder: optionSetOption
})
  .required()
  .label('contact-response')
