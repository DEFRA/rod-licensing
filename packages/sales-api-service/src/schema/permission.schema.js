import Joi from 'joi'
import { contactRequestSchema } from './contact.schema.js'
import { concessionProofSchema } from './concession-proof.schema.js'
import { optionSetOption } from './option-set.schema.js'
import { createReferenceDataEntityValidator } from './validators/validators.js'
import { Permit } from '@defra-fish/dynamics-lib'
import { validation } from '@defra-fish/business-rules-lib'
import { v4 as uuid } from 'uuid'

const issueDateSchema = Joi.string()
  .isoDate()
  .required()
  .allow(null)
  .description('An ISO8601 compatible date string defining when the permission was issued')
  .example(new Date().toISOString())
const startDateSchema = Joi.string()
  .isoDate()
  .required()
  .allow(null)
  .description('An ISO8601 compatible date string defining when the permission commences')
  .example(new Date().toISOString())
const endDateSchema = Joi.string()
  .isoDate()
  .required()
  .description('An ISO8601 compatible date string defining when the permission expires')
  .example(new Date().toISOString())

export const stagedPermissionSchema = Joi.object({
  permitId: Joi.string()
    .guid()
    .external(createReferenceDataEntityValidator(Permit))
    .required()
    .description('The ID of the permit associated with this permission')
    .example('cb1b34a0-0c66-e611-80dc-c4346bad0190'),
  licensee: contactRequestSchema,
  concessions: concessionProofSchema.optional(),
  issueDate: issueDateSchema.allow(null),
  startDate: startDateSchema.allow(null),
  isRenewal: Joi.boolean()
    .optional()
    .allow(null)
    .example('true')
}).label('staged-permission')

export const finalisedPermissionSchemaContent = {
  id: Joi.string()
    .guid()
    .required()
    .example(uuid()),
  referenceNumber: validation.permission.createPermissionNumberValidator(Joi),
  issueDate: issueDateSchema,
  startDate: startDateSchema,
  endDate: endDateSchema,
  stagingId: Joi.string()
    .guid()
    .required()
    .example(uuid()),
  dataSource: optionSetOption
}

export const finalisePermissionResponseSchema = stagedPermissionSchema
  .append({
    referenceNumber: validation.permission.createPermissionNumberValidator(Joi),
    endDate: endDateSchema
  })
  .label('finalised-permission')
