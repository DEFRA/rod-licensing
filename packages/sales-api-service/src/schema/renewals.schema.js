import Joi from 'joi'
import { validation } from '@defra-fish/business-rules-lib'
import { concessionProofSchema } from './concession-proof.schema.js'
import { permitSchema } from './permit.schema.js'
import { contactResponseSchema } from './contact.schema.js'
import { finalisedPermissionSchemaContent } from './permission.schema.js'

export const permissionRenewalDataRequestParamsSchema = Joi.object({
  referenceNumber: validation.permission.createPermissionNumberValidator(Joi)
}).label('permission-renewal-data-request-params')

export const permissionRenewalDataResponseSchema = Joi.object({
  permission: {
    ...finalisedPermissionSchemaContent,
    licensee: contactResponseSchema,
    concessions: concessionProofSchema,
    permit: permitSchema
  }
}).label('permission-renewal-data-response')
