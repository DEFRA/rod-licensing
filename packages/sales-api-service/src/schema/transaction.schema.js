import Joi from 'joi'
import { PoclFile } from '@defra-fish/dynamics-lib'
import { finalisePermissionResponseSchema, stagedPermissionSchema } from './permission.schema.js'
import { contactRequestSchema } from './contact.schema.js'
import { createAlternateKeyValidator, buildJoiOptionSetValidator, createPermitConcessionValidator } from './validators/validators.js'
import { MAX_PERMISSIONS_PER_TRANSACTION, POCL_TRANSACTION_SOURCES } from '@defra-fish/business-rules-lib'

import { v4 as uuidv4 } from 'uuid'

/**
 * Maximum number of items that can be created in a batch - limited by DynamoDB max batch size
 * @type {number}
 */
export const BATCH_CREATE_MAX_COUNT = 25

/**
 * Allow the existence of a transaction file to be cached for a period of time to reduce calls to Dynamics when processing POCL files
 * @type {number}
 */
export const TRANSACTION_FILE_VALIDATION_CACHE_TTL = 60 * 15 // 15 minutes

const createTransactionRequestSchemaContent = {
  permissions: Joi.array()
    .min(1)
    .max(MAX_PERMISSIONS_PER_TRANSACTION)
    .items(stagedPermissionSchema)
    .required()
    .label('create-transaction-request-permissions'),
  dataSource: buildJoiOptionSetValidator('defra_datasource', 'Web Sales'),
  serialNumber: Joi.when('dataSource', {
    is: Joi.valid(...POCL_TRANSACTION_SOURCES),
    then: Joi.string().trim().min(1).required()
  }),
  createdBy: Joi.string().optional(),
  journalId: Joi.string().optional()
}

/**
 * Schema for the create transaction request
 * @type {Joi.AnySchema}
 */
export const createTransactionSchema = Joi.object(createTransactionRequestSchemaContent)
  .external(createPermitConcessionValidator())
  .label('create-transaction-request')

/**
 * Validates a request to create transactions in batch.
 * Does not reference the createTransactionSchema as validation of objects in the array is applied manually in the request handler
 * so that a batch response can be formed (otherwise the entire batch will fail if any one of the create-requests contains a validation error)
 * @type {Joi.AnySchema}
 */
export const createTransactionBatchSchema = Joi.array()
  .min(1)
  .max(BATCH_CREATE_MAX_COUNT)
  .items(Joi.object().required().label('create-transaction-batch-item').description('See create-transaction-request for proper structure'))
  .required()
  .label('create-transaction-batch-request')

const createTransactionResponseSchemaContent = {
  id: Joi.string().trim().guid().required(),
  expires: Joi.number().required(),
  ...createTransactionRequestSchemaContent,
  permissions: Joi.array().min(1).items(stagedPermissionSchema).required().label('create-transaction-response-permissions'),
  cost: Joi.number().required(),
  isRecurringPaymentSupported: Joi.boolean().required(),
  status: Joi.object({
    id: Joi.string().valid('STAGED').required()
  })
    .label('create-transaction-status')
    .required()
}

/**
 * Schema for the create transaction response
 * @type {Joi.AnySchema}
 */
export const createTransactionResponseSchema = Joi.object(createTransactionResponseSchemaContent).label('create-transaction-response')

/**
 * Schema for the create transaction batch response
 * @type {Joi.AnySchema}
 */
export const createTransactionBatchResponseSchema = Joi.array()
  .items(
    Joi.object({
      statusCode: Joi.number().required(),
      response: Joi.object(createTransactionResponseSchemaContent).label('create-transaction-batch-item-response').optional(),
      error: Joi.string().optional(),
      message: Joi.string().optional()
    }).label('create-transaction-batch-response-item')
  )
  .label('create-transaction-batch-response')

const finaliseTransactionRequestSchemaContent = {
  transactionFile: Joi.string()
    .optional()
    .external(createAlternateKeyValidator(PoclFile, { cache: TRANSACTION_FILE_VALIDATION_CACHE_TTL })),
  payment: Joi.object({
    amount: Joi.number().required(),
    timestamp: Joi.string()
      .isoDate()
      .required()
      .description('An ISO8601 compatible date string defining when the transaction was completed')
      .example(new Date().toISOString()),
    source: buildJoiOptionSetValidator('defra_financialtransactionsource', 'Gov Pay'),
    channelId: Joi.string().trim().optional().description('Channel specific identifier'),
    method: buildJoiOptionSetValidator('defra_paymenttype', 'Debit card'),
    recurring: Joi.object({
      payer: contactRequestSchema,
      referenceNumber: Joi.string().required().description('The reference number associated with the recurring payment').example(uuidv4()),
      mandate: Joi.string().required().description('The mandate identifier associated with the recurring payment').example(uuidv4())
    })
      .label('finalise-transaction-recurring-payment-details')
      .description('Used to establish a recurring payment (e.g. via Direct Debit)')
      .optional()
  })
    .label('finalise-transaction-payment-details')
    .required()
}

/**
 * Schema for the finalise transaction request
 * @type {Joi.AnySchema}
 */
export const finaliseTransactionRequestSchema = Joi.object(finaliseTransactionRequestSchemaContent).label('finalise-transaction-request')

export const finaliseTransactionResponseSchema = Joi.object({
  ...createTransactionResponseSchemaContent,
  ...finaliseTransactionRequestSchemaContent,
  permissions: Joi.array().min(1).items(finalisePermissionResponseSchema).required().label('finalise-transaction-response-permissions'),
  status: Joi.object({
    id: Joi.string().valid('FINALISED').required(),
    messageId: Joi.string().required()
  })
    .required()
    .label('finalise-transaction-status')
}).label('finalise-transaction-response')
