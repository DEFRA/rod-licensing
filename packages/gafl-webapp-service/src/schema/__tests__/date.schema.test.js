import Joi from 'joi'
import { dateSchemaInput, dateSchema } from '../date.schema.js'

describe('dateSchemaInput', () => {
  it('matches expected format', () => {
    expect(dateSchemaInput('1', '2', '2023')).toMatchSnapshot()
  })
})

describe('dateSchema', () => {
  const getSamplePayload = ({ day = '', month = '', year = '' } = {}) => ({
    'date-of-birth-day': day,
    'date-of-birth-month': month,
    'date-of-birth-year': year
  })

  it.each`
    payload                                                     | expectedError          | payloadDesc
    ${{}}                                                       | ${'full-date'}         | ${'empty day, month and year'}
    ${{ year: '1' }}                                            | ${'day-and-month'}     | ${'empty day and month'}
    ${{ month: '2' }}                                           | ${'day-and-year'}      | ${'empty day and year'}
    ${{ day: '3' }}                                             | ${'month-and-year'}    | ${'empty month and year'}
    ${{ month: '5', year: '2023' }}                             | ${'day'}               | ${'empty day'}
    ${{ day: '12', year: '2024' }}                              | ${'month'}             | ${'empty month'}
    ${{ day: '15', month: '3' }}                                | ${'year'}              | ${'empty year'}
    ${{ day: 'Ides', month: 'March', year: '44 B.C.' }}         | ${'non-numeric.day'}   | ${'non-numerics entered'}
    ${{ day: 'Thirteenth', month: '11', year: '1978' }}         | ${'non-numeric.day'}   | ${'non-numeric day'}
    ${{ day: '29', month: 'MAR', year: '2002' }}                | ${'non-numeric.month'} | ${'non-numeric month '}
    ${{ day: '13', month: '1', year: 'Two thousand and five' }} | ${'non-numeric.year'}  | ${'non-numeric year'}
    ${{ day: '30', month: '2', year: '1994' }}                  | ${'invalid-date'}      | ${'an invalid date - 1994-02-40'}
    ${{ day: '1', month: '13', year: '2022' }}                  | ${'invalid-date'}      | ${'an invalid date - 2022-13-01'}
    ${{ day: '29', month: '2', year: '2023' }}                  | ${'invalid-date'}      | ${'an invalid date - 1994-02-40'}
  `('Error has $expectedError in details when payload has $payloadDesc', ({ payload: { day, month, year }, expectedError }) => {
    expect(() => {
      Joi.assert(dateSchemaInput(day, month, year), dateSchema)
    }).toThrow(
      expect.objectContaining({
        details: expect.arrayContaining([
          expect.objectContaining({
            path: expectedError.split('.'),
            context: expect.objectContaining({
              label: expectedError,
              key: expectedError.split('.').pop()
            })
          })
        ])
      })
    )
  })

  it('valid date passes validation', () => {
    expect(() => {
      Joi.assert(dateSchemaInput('12', '10', '1987'), dateSchema)
    }).not.toThrow()
  })
})
