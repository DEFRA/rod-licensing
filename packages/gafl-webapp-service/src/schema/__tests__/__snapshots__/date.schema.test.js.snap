// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`dateSchemaInput matches expected format 1`] = `
Object {
  "day": "1",
  "day-and-month": Object {
    "day": "1",
    "month": "2",
  },
  "day-and-year": Object {
    "day": "1",
    "year": "2023",
  },
  "full-date": Object {
    "day": "1",
    "month": "2",
    "year": "2023",
  },
  "invalid-date": "2023-02-01",
  "month": "2",
  "month-and-year": Object {
    "month": "2",
    "year": "2023",
  },
  "non-numeric": Object {
    "day": "1",
    "month": "2",
    "year": "2023",
  },
  "year": "2023",
}
`;
